python/unityagents/exception.py

```diff
@@ -10,3 +10,29 @@ class UnityActionException(Exception):
     Related to errors with sending actions.
     """
     pass
+
+class UnityTimeOutException(Exception):
+    """
+    Related to errors with communication timeouts.
+    """
+    def __init__(self, message, log_file_path = None):
+        if log_file_path is not None:
+            try:
+	            with open(log_file_path, "r") as f:
+	                printing = False
+	                for l in f:
+	                    l=l.strip()
+	                    if (l == 'Exception') or (l=='Error'):
+	                        printing = True
+	                        print('----------------------')
+	                    if (l == ''):
+	                        printing = False
+	                    if printing:
+	                        print(l)
+	                print("An error might have occured in the environment. "
+	                    "You can check the logfile for more information at {}".format(log_file_path))
+            except:
+                print("An error might have occured in the environment. "
```

Comment:
don't print, log. You can configure the logger at runtime to output to multiple places and formats including console or a log file without changing the code generating the messages thus decoupling generation from handling. Configure the logger using either dictConfig or fileConfig at runtime. Something like:

```python
logging.config.dictConfig({
  'handlers':{'file':{'class':'logging.FileHandler', 'filename':'unity-env.log'}
              'console':{'class':'logging.StreamHandler'}},
  'loggers':{'agents':{'handlers':['file', 'console']}}
})
logger = logging.getLogger('agents')
```

and then you can use the standard logger.info and logger.debug at will in the code and the messages will go to stderr and the log file you want and you can change that behavior by just changing the dict and not changing code

https://docs.python.org/3/library/logging.config.html#module-logging.config
https://docs.python.org/3/library/logging.handlers.html#logging.FileHandler

Reply:
For some reasons, I was not warned you requested changes here. I deleted the print statements in favor of logging. Thanks for the comment @romerocesar .